### DISCUSSION TITLE
Access Container from other Container does not work in Preview3

### DISCUSSION THREAD

**Discussion by nbauneau** [microsoft/aspire#2677] [Q&A] [answered]
I want to connect a custom container (in private ACR) to a RabbitMQ.

```cs
using Aspire.Hosting;

var builder = DistributedApplication.CreateBuilder(args);

var rabbitMq = builder.AddRabbitMQContainer2("messaging", port: 5672, password: "guest")
        .WithEnvironment("RabbitMQEventBus:UserName", "guest")
        .WithEnvironment("RabbitMQEventBus:Password", "guest")
        .WithEnvironment("RabbitMQEventBus:ExchangeName", "myexchange")
        .WithVolumeMount("../../deploy/DockerCompose/Local/RabbitMQ/docker-entrypoint.sh", "/usr/local/bin/docker-entrypoint.sh");

var my_custom_app= builder.AddContainer(name: "my_custom_app",
                                                image: "crhocustom.azurecr.io/my_custom_app",
                                                tag: "7.7.0-ft_feature")
        .WithEnvironment("RabbitMQEventBus__Connections__0__HostName", "host.docker.internal");
        .WithReference(rabbitMq)
        .WithEndpoint(containerPort: 80, hostPort: 5092, name: "endpoint", scheme: "http");

```
        
   

AddRabbitMQContainer2 methode :
```cs /// <summary>D
  /// Adds a RabbitMQ container to the application. The default image name is "rabbitmq" and the default tag is "3-management".
  /// </summary>
  /// <param name="builder">The <see cref="IDistributedApplicationBuilder"/>.</param>
  /// <param name="name">The name of the resource. This name will be used as the connection string name when referenced in a dependency.</param>
  /// <param name="port">The host port of RabbitMQ.</param>
  /// <param name="password">The password for RabbitMQ.</param>
  /// <returns>A reference to the <see cref="IResourceBuilder{T}"/>.</returns>
  public static IResourceBuilder<RabbitMQContainerResource> AddRabbitMQContainer2(this IDistributedApplicationBuilder builder, string name, int? port = null, string? password = null)
  {
      password ??= Guid.NewGuid().ToString("N");
      var rabbitMq = new RabbitMQContainerResource(name, password);
      return builder.AddResource(rabbitMq)
                     .WithAnnotation(new EndpointAnnotation(ProtocolType.Tcp, port: 5672, containerPort: 5672, name : "amqp"))// , isExternal: true
                     .WithAnnotation(new EndpointAnnotation(ProtocolType.Tcp, uriScheme: "http", name: "management", port: null, containerPort: 15672))
                     .WithAnnotation(new ContainerImageAnnotation { Image = "rabbitmq", Tag = "3-management" })
                     .WithManifestPublishingCallback(context => WriteRabbitMQContainerToManifest(context, rabbitMq))
                     .WithEnvironment("RABBITMQ_DEFAULT_USER", "guest")
                     .WithEnvironment(context =>
                     {
                         if (context.PublisherName == "manifest")
                         {
                             context.EnvironmentVariables.Add("RABBITMQ_DEFAULT_PASS", $"{{{rabbitMq.Name}.inputs.password}}");
                         }
                         else
                         {
                             context.EnvironmentVariables.Add("RABBITMQ_DEFAULT_PASS", rabbitMq.Password);
                         }
                     });

  }
```

- [x] AddProject is OK to connect to localhost:5672 
- [x] A console app in debug is ok to connect to host.docker.internal:5672 or localhost:5672 

- [ ] in AddContainer(name: "my_custom_app" i have : _RabbitMQ.Client.Exceptions.BrokerUnreachableException: None of the specified endpoints were reachable_

here is my RabbitMQ conf : 
![image](https://github.com/dotnet/aspire/assets/61270069/d00601f5-c064-4363-a192-0f82d8ad43ae)

can you help me ?

**Comment by davidfowl** [comment]
What do the logs say?

**Comment by nbauneau** [comment]
![image](https://github.com/dotnet/aspire/assets/61270069/3b8ca2e5-c8de-4595-99eb-6a03fd5b3555)
here are the logs of my custom containers
I have a retry policy for connecting to RabbitMQ

**Comment by davidfowl** [comment]
Can you print the address that the container is using in the logs?

**Comment by nbauneau** [comment]
Yes of course : 
![image](https://github.com/dotnet/aspire/assets/61270069/3dc65db1-1d3c-42a3-b25f-c52f2885940c)

**Comment by davidfowl** [comment]
It's unclear what might be happening, but there is a proxy forwarding tcp traffic to the container. I do recommend reading this https://learn.microsoft.com/en-us/dotnet/aspire/fundamentals/networking-overview.

**Comment by nbauneau** [comment]
the problem seems to be linked to the "bridge" mode and the fact that the containers are not in the same network (as one could do on docker-compose).

If I start the same container manually, I reproduce the same problem.
- ✖️ docker run mycustomcontainer ....
- ✅ docker run mycustomcontainer --link messaging-pa61nr8 ....

The --link command seems to be deprecated on docker and replaced by the use of network.

How to use a dedicated network or add the --link between AddContainer(...) on Aspire ?

**Comment by davidfowl** [comment]
There's no API support to go container to container in aspire. It's all through the host.

**Comment by nbauneau** [comment]
Thanks for the responsiveness 🙂
ok, but the containers cannot reach the "Aspire.Host" because it is exposed on 127.0.0.1 and we cannot change it to 0.0.0.0 of applicationUrl in launchSettings.json?

So container cannot use the services exposed by the proxy?

Another use case would be to be able to debug APIs used by a mobile application (Xamarin or Kotlin). This will be impossible as long as the App.Host remains accessible only from the workstation (localhost)

**Comment by DamianEdwards** [comment, marked as answer]
It can by substituting `localhost` for `host.docker.internal`

**Comment by nbauneau** [comment]
Yes, in effect. You are right @DamianEdwards  and @davidfowl !
Sorry, i haven't specified that I used rancher desktop to develop containers on Windows...
This usage of host.docker.internal from containers (between containers from different networks)  is only available with this enabled option : https://docs.rancherdesktop.io/ui/preferences/wsl/network/
I thought that is an Aspire Bug, but it's only a Rancher Desktop configuration 🫤

=> however, I still have the problem exposing the services started by Aspire on the IP of my workstation

In any case, great project! THANKS 💪🚀
